File: cpp_engine/portfolio.cpp
#include "portfolio.hpp"

#include <stdexcept>

namespace quant {

Portfolio::Portfolio(double initial_cash) : initial_cash_(initial_cash), cash_(initial_cash) {
    if (initial_cash <= 0.0) {
        throw std::invalid_argument("initial_cash must be positive");
    }
}

double Portfolio::cash() const {
    return cash_;
}

int Portfolio::quantity(const std::string& ticker) const {
    const auto it = positions_.find(ticker);
    return it == positions_.end() ? 0 : it->second.quantity();
}

double Portfolio::holdings_value(const std::unordered_map<std::string, double>& prices) const {
    double value = 0.0;
    for (const auto& item : positions_) {
        const auto price_it = prices.find(item.first);
        if (price_it != prices.end()) {
            value += item.second.market_value(price_it->second);
        }
    }
    return value;
}

double Portfolio::total_value(const std::unordered_map<std::string, double>& prices) const {
    return cash_ + holdings_value(prices);
}

const std::vector<TradeRecord>& Portfolio::trades() const {
    return trades_;
}

void Portfolio::apply_fill(const Fill& fill) {
    auto& position = positions_.try_emplace(fill.ticker, fill.ticker).first->second;
    const double gross = fill.notional();

    if (fill.side == OrderSide::Buy) {
        const double total_cost = gross + fill.transaction_cost;
        if (total_cost > cash_ + 1e-9) {
            throw std::runtime_error("insufficient cash");
        }
        position.buy(fill.quantity, fill.executed_price);
        cash_ -= total_cost;
    } else {
        position.sell(fill.quantity);
        cash_ += gross - fill.transaction_cost;
        if (position.quantity() == 0) {
            positions_.erase(fill.ticker);
        }
    }

    trades_.push_back(
        TradeRecord{
            fill.date,
            fill.ticker,
            fill.side,
            fill.quantity,
            fill.executed_price,
            fill.transaction_cost,
            cash_,
        }
    );
}

}  // namespace quant

File: src/backtesting/portfolio.py
"""Portfolio accounting for long-only equity backtests."""

from __future__ import annotations

from dataclasses import asdict, dataclass

from backtesting.execution import Fill


@dataclass
class Position:
    """Single-name position state."""

    ticker: str
    quantity: int = 0
    avg_cost: float = 0.0

    def market_value(self, price: float) -> float:
        """Return current marked value."""

        return self.quantity * price


class Portfolio:
    """Cash, positions, and trade ledger."""

    def __init__(self, initial_cash: float = 100_000.0) -> None:
        if initial_cash <= 0:
            raise ValueError("initial_cash must be positive")
        self.initial_cash = float(initial_cash)
        self.cash = float(initial_cash)
        self.positions: dict[str, Position] = {}
        self.trade_history: list[dict[str, object]] = []

    def quantity(self, ticker: str) -> int:
        """Return held share quantity for a ticker."""

        position = self.positions.get(ticker)
        return 0 if position is None else position.quantity

    def holdings_value(self, prices: dict[str, float]) -> float:
        """Mark holdings to supplied prices."""

        return sum(position.market_value(prices.get(ticker, 0.0)) for ticker, position in self.positions.items())

    def total_value(self, prices: dict[str, float]) -> float:
        """Return cash plus marked holdings."""

        return self.cash + self.holdings_value(prices)

    def execute_fill(self, fill: Fill) -> None:
        """Apply a fill to cash, position state, and trade history."""

        ticker = fill.ticker
        position = self.positions.get(ticker, Position(ticker=ticker))

        if fill.side == "BUY":
            total_cost = fill.notional + fill.transaction_cost
            if total_cost > self.cash + 1e-8:
                raise ValueError(f"insufficient cash for {ticker} buy")
            new_quantity = position.quantity + fill.quantity
            if new_quantity <= 0:
                raise ValueError("buy produced non-positive position")
            position.avg_cost = (
                position.avg_cost * position.quantity + fill.executed_price * fill.quantity
            ) / new_quantity
            position.quantity = new_quantity
            self.cash -= total_cost
        else:
            if fill.quantity > position.quantity:
                raise ValueError(f"cannot sell more {ticker} shares than held")
            position.quantity -= fill.quantity
            self.cash += fill.notional - fill.transaction_cost
            if position.quantity == 0:
                position.avg_cost = 0.0

        if position.quantity:
            self.positions[ticker] = position
        else:
            self.positions.pop(ticker, None)

        row = asdict(fill)
        row["cash_after"] = self.cash
        self.trade_history.append(row)

    def position_snapshot(self, date: object, prices: dict[str, float]) -> list[dict[str, object]]:
        """Create position rows for reporting."""

        rows = []
        for ticker, position in sorted(self.positions.items()):
            price = prices.get(ticker, 0.0)
            rows.append(
                {
                    "date": date,
                    "ticker": ticker,
                    "quantity": position.quantity,
                    "avg_cost": position.avg_cost,
                    "market_price": price,
                    "market_value": position.market_value(price),
                }
            )
        return rows
